### PR TITLE
Ben/float 16

### PR DESCRIPTION
**Pull request by benleetownsend**

### CODE REVIEW

File: tests/test_base_models.py [hunk 1]
```diff
@@ -241,6 +241,23 @@ def test_save_load(self):
...
 
+    def test_fp_16_equality(self):
+        config = self.default_config(save_adam_vars=False, float_16_predict=True)
+        model = Classifier(**config)
+        train_sample = self.dataset.sample(n=self.n_sample)
+        valid_sample = self.dataset.sample(n=self.n_sample)
+        model.fit(train_sample.Text, train_sample.Target)
+        predictions = model.predict(valid_sample.Text)
+        model.config.float_16_predict = False
+        # important to do it this way for the models that do not support fp16 as
```
Comment (madisonmay) [inline tests/test_base_models.py:252]: Don't quite follow this comment. Could you explain further?

File: tests/test_base_models.py [hunk 2]
```diff
@@ -241,6 +241,23 @@ def test_save_load(self):
...
+        # config resolution is only done when the config is set correctly.
+        new_predictions = model.predict(valid_sample.Text)
+        errors = 0
+        for prediction, new_prediction in zip(predictions, new_predictions):
+            if prediction != new_prediction:
```
Comment (madisonmay) [inline tests/test_base_models.py:257]: Exact equality works here?!

File: tests/test_parity.py [hunk 3]
```diff
@@ -90,3 +90,30 @@ def test_roberta_featurize(self):
...
+        model = Classifier(base_model=BERT, float_16_predict=True)
+        np.testing.assert_allclose(
+            model.featurize(self.TEST_DATA)[0],
+            np.load(
+                os.path.join(
+                    DIRECTORY,
+                    'data/test-bert-activations.npy'
+                )
+            ),
+            atol=1e-1
```
Comment (lhz1029) [inline tests/test_parity.py:104]: That tolerance seems quite high. Is that what we expect?

File: tests/test_base_models.py (hunk 2 above)
Comment (lhz1029) [reply tests/test_base_models.py:257]: Yeah I'm quite surprised too

**Review by lhz1029 — APPROVED** [review]
For the most part LGTM! Out of curiosity, how did you go through the codebase to make sure you got everything? Just a search for all tf.get_variable, tf.cast, tf.zeros, and tf.ones?

File: tests/test_base_models.py (hunk 2 above)
Comment (benleetownsend) [reply tests/test_base_models.py:257]: Over a large enough dataset i'm sure a classification or two will come out differently, and for sequence labelling I have seen results vary by a token either way.

File: tests/test_parity.py (hunk 3 above)
Comment (benleetownsend) [reply tests/test_parity.py:104]: It's the same tolerance we have for the main BERT/Roberta Models, but yes it is pretty high.

**Comment by benleetownsend** [comment]
@lhz1029 I changed all the variables with the overridden variable getter because we don't always have direct access to the variables themselves if some layers or keras module is used and then just searched for tf.float32 to find any explicit casts and changed these to cast to the dtype that it needs to be compatible with. Then set the default config to use float_16 on every inference call and ran the whole test suite to see where any implicit float32s were coming from.

File: tests/test_base_models.py (hunk 1 above)
Comment (benleetownsend) [reply tests/test_base_models.py:252]: So when we explicitly set model.config.* it doesn't go through the config resolution step like if you set the config items properly. Only BERT and RoBERTa support the float_16 predict but in these cases a warning is raised and this config argument is set back to False. If we start with False and then use `model.config.float_16_predict = True` on any of the base models where it is not supported this will fail because the config resolution doesn't catch it.

Comment (benleetownsend) [reply tests/test_base_models.py:252]: So the real question I guess is why do we test the models that don't support it yet, but I think it's important to get defined behaviour in these cases so that someone with a GPU that supports it can just set this config argument for every model and not worry about getting undefined behaviour.

Another option would be to explicitly raise an exception in these cases and then change the tests accordingly.

We should maybe think about moving the config resolution into the config object itself though so that any time a conflicting set of config arguments are set this is resolved no matter how or when they are set.

File: finetune/model.py [hunk 4]
```diff
@@ -132,11 +142,17 @@ def _model_fn(features, labels, mode, params):
         train = estimator_mode == tf.estimator.ModeKeys.TRAIN
         X = features["tokens"]
         context = features.get("context", None)
-        task_id = features.get("task_id", None)
         Y = labels
         pred_op = None
 
-        with tf.variable_scope(tf.get_variable_scope()):
+        if estimator_mode == tf.estimator.ModeKeys.PREDICT and params.float_16_predict:
```
Comment (madisonmay) [inline finetune/model.py]: Only request would be factoring this out into a separate fn to keep _model_fn relatively clean

File: tests/test_base_models.py (hunk 2 above)
Comment (madisonmay) [reply tests/test_base_models.py:257]: Ah I was thinking about probas.  But these are just ints / labels

**Review by madisonmay — APPROVED** [review]

File: tests/test_base_models.py (hunk 1 above)
Comment (lhz1029) [reply tests/test_base_models.py:252]: yeah, moving to the config object seems like the cleanest approach.

File: finetune/base.py [hunk 5]
```diff
@@ -317,6 +318,12 @@ def get_estimator(self, force_build_lm=False, build_explain=False, cache=False):
         else:
             build_lm = force_build_lm or self.config.lm_loss_coef > 0.0
             config = self._get_estimator_config()
+            
+            fp16_predict = self.config.float_16_predict
+            if fp16_predict:
+                if not gpu_info(config.session_config)["fp16_inference"]:
+                    fp16_predict = False
```
Comment (lhz1029) [inline finetune/base.py:328]: maybe add a warning here to the user?

File: finetune/util/gpu_info.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        if local_device.device_type == "GPU":
+            num_gpus += 1
+            fp16 = is_fp16_capable(local_device.physical_device_desc)
+            if fp16 and not fp16_capable:
+                if num_gpus == 1:
+                    fp16_capable = True
+                else:
+                    LOGGER.warning(
+                        (
+                            "A GPU with float 16 support is available but it is not GPU:0,"
```
Comment (lhz1029) [inline finetune/util/gpu_info.py:26]: What happens in the case of multiple gpus, some with float 16 support and others without?

Comment (lhz1029) [reply finetune/util/gpu_info.py:26]: If it's too difficult to manage the various cases, we could just not support float16 unless all gpus that are visible have float16 enabled. Is that possible/easy to do?

**Review by lhz1029 — APPROVED** [review]

File: finetune/config.py [hunk 7]
```diff
@@ -239,7 +239,7 @@ def get_default_config():
     settings = Settings(
         # General Settings
         low_memory_mode=False,
-        float_16_predict=False,
+        float_16_predict=True,
```
Comment (lhz1029) [inline finetune/config.py]: oh we've decided to make this the default?

File: finetune/util/gpu_info.py (hunk 6 above)
Comment (benleetownsend) [reply finetune/util/gpu_info.py:26]: By default predict is put into the first gpu. So if this is not float 16, we don't use float 16, this warning is just to let people know that.

File: finetune/base.py (hunk 5 above)
Comment (madisonmay) [reply finetune/base.py:328]: yeah would just be a matter of changing this to LOGGER.warning

File: finetune/config.py (hunk 7 above)
Comment (madisonmay) [reply finetune/config.py]: I think let's make it opt in for a bit and then change the default once we've built some trust that it won't ever do us harm?  Have no reason to believe this should ever be problematic though given that the gpu_info check is in place.

File: finetune/model.py [hunk 8]
```diff
@@ -123,6 +163,7 @@ def target_model_op(featurizer_state, Y, params, mode, **kwargs):
         return target_model_state
 
     def _model_fn(features, labels, mode, params):
+        var_getter, features = get_variable_getter(mode, features, fp16_predict)
```
Comment (madisonmay) [inline finetune/model.py:166]: Sweet, thanks!

**Comment by benleetownsend** [comment]
All tests pass.

File: finetune/util/gpu_info.py (hunk 6 above)
Comment (lhz1029) [reply finetune/util/gpu_info.py:26]: :+1:
